src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -865,65 +865,73 @@ private int ReadBuffer(Span<char> userBuffer, out bool readToUserBuffer)
                 do
                 {
-                    char ch = tmpCharBuffer[i];
+                    char[] tmpCharBuffer = _charBuffer;
+                    int tmpCharLen = _charLen;
+                    int tmpCharPos = _charPos;
+                    int i = tmpCharPos;
 
-                    // Note the following common line feed chars:
-                    // \n - UNIX   \r\n - DOS   \r - Mac
-                    if (ch == '\r' || ch == '\n')
+                    do
                     {
-                        string s;
+                        char ch = tmpCharBuffer[i];
 
-                        if (sb != null)
-                        {
-                            sb.Append(tmpCharBuffer, tmpCharPos, i - tmpCharPos);
-                            s = sb.ToString();
-                        }
-                        else
+                        // Note the following common line feed chars:
+                        // \n - UNIX   \r\n - DOS   \r - Mac
+                        if (ch is '\r' or '\n')
                         {
-                            s = new string(tmpCharBuffer, tmpCharPos, i - tmpCharPos);
-                        }
+                            string s;
 
-                        _charPos = tmpCharPos = i + 1;
+                            if (sb != null)
+                            {
+                                sb.Append(tmpCharBuffer, tmpCharPos, i - tmpCharPos);
+                                s = sb.ToString();
+                            }
+                            else
+                            {
+                                s = new string(tmpCharBuffer, tmpCharPos, i - tmpCharPos);
+                            }
 
-                        if (ch == '\r' && (tmpCharPos < tmpCharLen || (await ReadBufferAsync(CancellationToken.None).ConfigureAwait(false)) > 0))
-                        {
-                            tmpCharPos = _charPos;
-                            if (_charBuffer[tmpCharPos] == '\n')
+                            _charPos = tmpCharPos = i + 1;
+
+                            if (ch == '\r' && (tmpCharPos < tmpCharLen || (await ReadBufferAsync(CancellationToken.None).ConfigureAwait(false)) > 0))
                             {
-                                _charPos = ++tmpCharPos;
+                                tmpCharPos = _charPos;
+                                if (_charBuffer[tmpCharPos] == '\n')
+                                {
+                                    _charPos = ++tmpCharPos;
+                                }
                             }
+
+                            return s;
                         }
 
-                        return s;
-                    }
+                        i++;
+                    } while (i < tmpCharLen);
 
-                    i++;
-                } while (i < tmpCharLen);
+                    i = tmpCharLen - tmpCharPos;
+                    sb ??= new StringBuilder(i + 80);
```

Comment:
@adamsitnik I was wondering whether it would be possible to do pooling of the `StringBuilder` instance, to avoid this allocation in long line cases?

Reply:
@nietras it's definitely a good idea! Perhaps we could even replace the private char array ( `_charBuffer`) with a `StringBuilder`, write the chars directly to it and avoid the need of appending? The only thing I am not sure of is what side effects using `StringBuilder.ToString(int startIndex, int length)` have